### PR TITLE
Change SHOW COLUMNS query to display MySQL types in MySQL Compatibility mode

### LINKED ISSUE #49168: MySQL compatibility: native type aliases
**Use case**

Currently, while using MySQL protocol, `SHOW COLUMNS` or `SELECT ... FROM system.columns` returns ClickHouse types instead of MySQL, and this breaks the type introspection for BI tools.

For example:

```
SHOW FULL COLUMNS FROM `cell_towers` FROM `default` LIKE '%';

+---------------+-----------------------------------------------------------------------+------+---------+---------+-------+
| field         | type                                                                  | null | key     | default | extra |
+---------------+-----------------------------------------------------------------------+------+---------+---------+-------+
| area          | UInt16                                                                |    0 |         | NULL    |       |
| averageSignal | UInt8                                                                 |    0 |         | NULL    |       |
| cell          | UInt64                                                                |    0 |         | NULL    |       |
| changeable    | UInt8                                                                 |    0 |         | NULL    |       |
| created       | DateTime                                                              |    0 | PRI SOR | NULL    |       |
| lat           | Float64                                                               |    0 |         | NULL    |       |
| lon           | Float64                                                               |    0 |         | NULL    |       |
| mcc           | UInt16                                                                |    0 | PRI SOR | NULL    |       |
| net           | UInt16                                                                |    0 | PRI SOR | NULL    |       |
| radio         | Enum8('' = 0, 'CDMA' = 1, 'GSM' = 2, 'LTE' = 3, 'NR' = 4, 'UMTS' = 5) |    0 | PRI SOR | NULL    |       |
| range         | UInt32                                                                |    0 |         | NULL    |       |
| samples       | UInt32                                                                |    0 |         | NULL    |       |
| unit          | Int16                                                                 |    0 |         | NULL    |       |
| updated       | DateTime                                                              |    0 |         | NULL    |       |
+---------------+-----------------------------------------------------------------------+------+---------+---------+-------+
```

**Describe the solution you'd like**

Alias ClickHouse type to MySQL native types via MySQL protocol, i.e., `Boolean` = `TINYINT`, `String` = `VARCHAR`, etc. The preferred way is to have it enabled by default via MySQL protocol, but a server configuration setting will do if it's not possible.

**Additional context**

This is the main problem with QuickSight when accessing ClickHouse via MySQL protocol.

For example, the [`cell_towers`](https://clickhouse.com/docs/en/getting-started/example-datasets/cell-towers) data types introspection failed; all the fields except `DateTime` ones (`created` and `updated`, cause MySQL has `DateTime` type natively) were marked as `OTHER` and excluded from the dataset.

image

CC @rschu1ze

### PR DESCRIPTION
**Pull request by tpanetti**
This updates the SHOW COLUMN SQL query to display MySQL types when this query is issued by a client connected via MySQL Compatibility port

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a new setting named `use_mysql_types_in_show_columns` to alter the `SHOW COLUMNS` SQL statement to display MySQL equivalent types when a client is connected via the MySQL compatibility port. 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

Motivation: This functionality is useful for ClickHouse users as it allows tools that use standard MySQL connectors to read and parse ClickHouse data as if it were MySQL. This is particularly useful for data analytics tools such as QuickSights and Looker Studio.
Parameters: No arguments or parameters are needed, this type conversion is done when a client is connected via the MySQL port. Specifically, the implementation happens when the client is connected to whatever port the configuration has set for MySQL compatibility.
Example use:
`SHOW FULL COLUMNS`

I've attached an image of Amazon QuickSight parsing a ClickHouse database with example cell_towers data with this code change.
> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/
Screenshot 2023-05-05 at 12 13 27 PM

Closes [#49168](https://github.com/ClickHouse/ClickHouse/issues/49168)

### CODE REVIEW

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 1]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
 
 }
 
+String InterpreterShowColumnsQuery::getMySQLQuery()
+{
+    WriteBufferFromOwnString mysql_specific_query;
+
+    mysql_specific_query << "SELECT name AS field, "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:106]: It does not look maintainable.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 2]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "CASE "
+        << "  WHEN startsWith(type, 'Nullable') THEN "
+        << "    CASE "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt8', 'Int8') THEN 'tinyint' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:110]: The code does not explain what 10 is.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 3]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt16', 'Int16') THEN 'smallint' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:111]: This introduces an implicit dependency between this file and every file inside the DataTypes directory.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 4]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt32', 'Int32') THEN 'int' "
+        << "      WHEN substring(type, 10, length(type) - 10) IN ('UInt64', 'Int64', 'UInt128', 'Int128', 'UInt256', 'Int256') THEN 'bigint' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:113]: This is wrong because MySQL's bigint is just 64 bits.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 5]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) = 'Float32' THEN 'float' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'Float64' THEN 'double' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Decimal%' THEN 'decimal' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'Boolean' THEN 'tinyint' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'String' THEN 'text' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:118]: String can contain arbitrary binary data.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 6]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'FixedString%' THEN 'text' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Date%' THEN 'date' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'DateTime%' THEN 'datetime' "
+        << "      WHEN substring(type, 10, length(type) - 10) = 'JSON' THEN 'json' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:122]: This does not look correct.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 7]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) = 'UUID' THEN 'binary' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:123]: Why binary?

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 8]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Enum%' THEN 'enum' "
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'LowCardinality%' THEN 'text' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:125]: This is wrong.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp [hunk 9]
```diff
@@ -93,6 +99,74 @@ String InterpreterShowColumnsQuery::getRewrittenQuery()
...
+        << "      WHEN substring(type, 10, length(type) - 10) LIKE 'Array%' THEN 'json' "
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterShowColumnsQuery.cpp:126]: I'm not sure that this will work.

File: src/Interpreters/InterpreterShowColumnsQuery.cpp (hunk 3 above)
Comment (tpanetti) [reply src/Interpreters/InterpreterShowColumnsQuery.cpp:111]: This is a good point. Perhaps we should instead build this string by iterating over the TypeIndex enum for all of the ClickHouse types and get the equivalent MySQL type with a function getSQLCompatibleName in IDataType (as you mentioned before), then look it up in a type equivalency map. That should solve the implicit dependency and make this more manageable although we would need to maintain the equivalency map.

**Comment by tpanetti** [comment]
Refactored to a more scalable method of implementation via the IDataType interface. Added @rschu1ze as an additional reviewer after discussions with him about this change. For incompatible types (eg int64, int128) I chose to default to text for compatibility with BI tools, this may warrant further discussion though.

**Comment by alexey-milovidov** [comment]
Ok. Need to fix the test. Copy-paste the initialization header from similar .sh tests.

**Comment by alexey-milovidov** [comment]
Let's rename `getMySQLName` to `getSQLCompatibleName`, and change the returned value to ALL CAPS (`TEXT`, etc).

**Comment by tpanetti** [comment]
Added the initialization header from similar .sh tests, fixed the casing of the MySQL types and renamed `getMySQLName` to `getSQLCompatibleName`

File: src/DataTypes/DataTypeEnum.h [hunk 10]
```diff
@@ -45,13 +45,16 @@ class DataTypeEnum final : public IDataTypeEnum, public EnumValues<Type>
...
+    std::string my_sql_type_name;
     static std::string generateName(const Values & values);
+    static std::string generateMySQLName(const Values & values);
 
 public:
     explicit DataTypeEnum(const Values & values_);
 
     std::string doGetName() const override { return type_name; }
     const char * getFamilyName() const override;
+    const char * getSQLCompatibleName() const override { return my_sql_type_name.c_str(); }
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeEnum.h:57]: Now returning const char * starts to be dangerous, let's change the interface of this function to return String.

Also, let's generate it on the fly - do not store it in the class's member.

File: src/DataTypes/DataTypeLowCardinality.h [hunk 11]
```diff
@@ -22,6 +24,8 @@ class DataTypeLowCardinality : public IDataType
         return "LowCardinality(" + dictionary_type->getName() + ")";
     }
     const char * getFamilyName() const override { return "LowCardinality"; }
+    const char * getSQLCompatibleName() const override { return mysql_name.c_str(); }
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeLowCardinality.h:27]: .

File: src/DataTypes/DataTypeNumberBase.h [hunk 12]
```diff
@@ -20,11 +20,14 @@ class DataTypeNumberBase : public IDataType
...
     static constexpr auto family_name = TypeName<T>;
     static constexpr auto type_id = TypeToTypeIndex<T>;
+    // Create a map from the name of the type to the name of the type in MySQL.
+    static const std::map<std::string, std::string> mysqlTypeMap;
 
     using FieldType = T;
     using ColumnType = ColumnVector<T>;
 
     const char * getFamilyName() const override { return TypeName<T>.data(); }
+    const char * getSQLCompatibleName() const override { return mysqlTypeMap.at(TypeName<T>.data()).c_str(); }
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeNumberBase.h:30]: We should avoid the usage of TypeName. I think writing a series of `if constexpr` directly here will be better.

File: src/DataTypes/DataTypeString.h [hunk 13]
```diff
@@ -21,6 +21,9 @@ class DataTypeString final : public IDataType
         return "String";
     }
 
+    // FIXME: string can contain arbitrary bytes, not only UTF-8 sequences
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeString.h:24]: Nothing to fix here.

But it raises a question, why we are using TEXT for AggregateFunction - it contains non-UTF8 sequences almost always.

File: src/DataTypes/DataTypeTuple.h [hunk 14]
```diff
@@ -33,6 +33,7 @@ class DataTypeTuple final : public IDataType
     TypeIndex getTypeId() const override { return TypeIndex::Tuple; }
     std::string doGetName() const override;
     const char * getFamilyName() const override { return "Tuple"; }
+    const char * getSQLCompatibleName() const override { return "JSON"; }
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeTuple.h:36]: Not sure how external tools will eat it when we provide values of this type.

File: src/DataTypes/DataTypeUUID.h [hunk 15]
```diff
@@ -18,6 +18,8 @@ class DataTypeUUID : public IDataType
     static constexpr auto type_id = TypeIndex::UUID;
 
     const char * getFamilyName() const override { return "UUID"; }
+    const char * getSQLCompatibleName() const override { return "CHAR"; }
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeUUID.h:21]: CHAR in MySQL is of fixed size. Specifying CHAR without a size can lead to something strange.

Comment (alexey-milovidov) [reply src/DataTypes/DataTypeUUID.h:21]: Although it might work...

File: src/DataTypes/IDataType.h [hunk 16]
```diff
@@ -71,10 +71,19 @@ class IDataType : private boost::noncopyable, public std::enable_shared_from_thi
           return doGetName();
     }
 
+    /// MySQL equivalent Name of data type (examples: UInt64, Array(String)).
+    String getMySQLTypeName() const
+    {
+        if (custom_name)
```
Comment (alexey-milovidov) [inline src/DataTypes/IDataType.h:77]: I don't think that this branch is relevant.

File: src/Storages/System/StorageSystemColumns.cpp [hunk 17]
```diff
@@ -74,6 +74,7 @@ class ColumnsSource : public ISource
         : ISource(header_)
         , columns_mask(std::move(columns_mask_)), max_block_size(max_block_size_)
         , databases(std::move(databases_)), tables(std::move(tables_)), storages(std::move(storages_))
+        , clientInfo(context->getClientInfo())
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemColumns.cpp:77]: clientInfo -> client_info,
but it is better not to copy the whole struct here but only check for a particular member and set a flag.

File: src/Storages/System/StorageSystemColumns.cpp [hunk 18]
```diff
@@ -129,6 +130,17 @@ class ColumnsSource : public ISource
 
             bool check_access_for_columns = check_access_for_tables && !access->isGranted(AccessType::SHOW_COLUMNS, database_name, table_name);
 
+            auto get_type_name = [this](const IDataType& type) -> std::string
+            {
+                if (clientInfo.interface == DB::ClientInfo::Interface::MYSQL)
+                {
+                   return type.getMySQLTypeName();
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemColumns.cpp:137]: Wrong indent - three spaces instead of four.
Wrong function name.

**Comment by alexey-milovidov** [comment]
If we change the behavior of the MySQL interface by default, it is going to be a backward incompatible change - let's make a setting to control its behavior.

File: src/DataTypes/DataTypeFixedString.h [hunk 19]
```diff
@@ -42,6 +42,7 @@ class DataTypeFixedString final : public IDataType
     TypeIndex getTypeId() const override { return type_id; }
 
     const char * getFamilyName() const override { return "FixedString"; }
+    const char * getSQLCompatibleName() const override { return "TEXT"; }
```
Comment (rschu1ze) [inline src/DataTypes/DataTypeFixedString.h:45]: Isn't that `CHAR` in MySQL? https://dev.mysql.com/doc/refman/8.0/en/char.html

EDIT: Ah, guess TEXT is more fitting because it can hold arbitrary bytes and CHAR can't. Maybe add a comment like in DataTypeString.h.

File: src/DataTypes/DataTypesDecimal.h [hunk 20]
```diff
@@ -37,8 +37,11 @@ class DataTypeDecimal final : public DataTypeDecimalBase<T>
     using Base::Base;
 
     static constexpr auto family_name = "Decimal";
+    static constexpr auto mysql_name = "DECIMAL";
 
     const char * getFamilyName() const override { return family_name; }
+    const char * getSQLCompatibleName() const override { return mysql_name; }
```
Comment (rschu1ze) [inline src/DataTypes/DataTypesDecimal.h:43]: It will be cleaner to include `M` and `D` parameters: https://dev.mysql.com/doc/refman/8.0/en/fixed-point-types.html

File: tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh [hunk 21]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env bash
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:1]: Curious why this test can't be a "normal" .sql test?

File: tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+${CLICKHOUSE_LOCAL} --query "DROP TABLE IF EXISTS database_123456789abcde.tab"
+
+#${CLICKHOUSE_LOCAL} --query "SET allow_suspicious_low_cardinality_types = 1;"
+echo "Create tab table "
+${CLICKHOUSE_LOCAL} -n -q "
+    SET allow_suspicious_low_cardinality_types=1;
+    SET allow_experimental_object_type =1;
+    CREATE TABLE tab
+    (
+        uint64 UInt64,
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:27]: (Nitpicking: not all integer types are tested).

File: tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh (hunk 21 above)
Comment (tpanetti) [reply tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:1]: Since we're testing both the clickhouse-client and mysql client simultaneously we need to use a shell script to leverage both tools.

**Comment by alexey-milovidov** [comment]
@rschu1ze VARCHAR/TEXT/BLOB, etc are SQL standard types, and we are mostly inside this set.

**Comment by tpanetti** [comment]
Addressed all comments, added the configuration setting named `output_format_mysql_types` to enable MySQL type conversion so we can drop the `pr-backward-incompatible` tag.

File: src/DataTypes/IDataType.h [hunk 23]
```diff
@@ -75,6 +75,8 @@ class IDataType : private boost::noncopyable, public std::enable_shared_from_thi
 
     /// Name of data type family (example: FixedString, Array).
     virtual const char * getFamilyName() const = 0;
+    /// Name of corresponding data type in MySQL (exampe: Bigint, Blob, etc)
```
Comment (rschu1ze) [inline src/DataTypes/IDataType.h:78]: (For consistency, it would be great to have `getFamilyName()` also return a String. But that's unrelated and fine to be done in a separate PR.)

File: src/DataTypes/DataTypeNumberBase.cpp [hunk 24]
```diff
@@ -11,6 +11,55 @@ Field DataTypeNumberBase<T>::getDefault() const
 {
     return NearestFieldType<FieldType>();
 }
+template <typename T>
+String DataTypeNumberBase<T>::getSQLCompatibleName() const
+{
+    if constexpr (std::is_same_v<T, Int8>)
```
Comment (rschu1ze) [inline src/DataTypes/DataTypeNumberBase.cpp:17]: (I don't care too much but parentheses for single-line if/for/while statements are usually omitted in the codebase. In this case, parentheses blow the code artificially up.)

**Review by rschu1ze — COMMENTED** [review]
Lgtm, thanks.

File: src/DataTypes/IDataType.h (hunk 23 above)
Comment (alexey-milovidov) [reply src/DataTypes/IDataType.h:78]: It's ok for `getFamilyName` to return a reference to a statically allocated string (const char *), because the family name is constant for the class.

File: src/Core/Settings.h [hunk 25]
```diff
@@ -196,6 +196,7 @@ class IColumn;
     M(Bool, allow_experimental_inverted_index, false, "If it is set to true, allow to use experimental inverted index.", 0) \
     \
     M(UInt64, mysql_max_rows_to_insert, 65536, "The maximum number of rows in MySQL batch insertion of the MySQL storage engine", 0) \
+    M(Bool, output_format_mysql_types, false, "Use MySQL converted types when connected via MySQL compatibility", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:199]: All LGTM except this setting name and description.

It does not force output MySQL types everywhere, only for SHOW COLUMNS. For example, it does not affect Native, JSON, and CSVWithNamesAndTypes output formats.

Let's rename it to `use_mysql_types_in_show_columns`?

**Review by rschu1ze — APPROVED** [review]
Lgtm (thanks) but the tests still fail in a weird way.

E.g. here: https://s3.amazonaws.com/clickhouse-test-reports/49577/cb8c20722b8976fe0bc402498667b02c2585cc02/stateless_tests_flaky_check__asan_.html

I did not check deeper but there can be a race on database name `database_123456789abcde` when the test runs in parallel. You could use tag "no-parallel" at the top of the test file to suppress parallel execution.

File: tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+USER="default"
+PASSWORD=""
+HOST="127.0.0.1"
+PORT=9004
+
+# First run the clickhouse test to create the ClickHouse Tables
+
+echo "Drop tables if they exist"
+${CLICKHOUSE_LOCAL} --query "DROP TABLE IF EXISTS tab"
+${CLICKHOUSE_LOCAL} --query "DROP TABLE IF EXISTS database_123456789abcde"
```
Comment (Avogar) [inline tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:18]:
```suggestion
${CLICKHOUSE_LOCAL} --query "DROP DATABASE IF EXISTS database_123456789abcde"
```

Comment (tpanetti) [reply tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh]: Fixed

**Comment by rschu1ze** [comment]
ClickHouse Integration Tests (asan) [6/6]:
- test_ssl_cert_authentication: #49400

ClickHouse Stateless Tests (debug) [2/5]
- 02703_max_local_write_bandwidth #51221
- 02561_null_as_default_more_formats #49177

File: tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+USER="default"
+PASSWORD=""
+HOST="127.0.0.1"
+PORT=9004
+
+# First run the clickhouse test to create the ClickHouse Tables
+
+echo "Drop tables if they exist"
+${CLICKHOUSE_CLIENT} --query "DROP TABLE IF EXISTS tab"
+${CLICKHOUSE_CLIENT} --query "DROP DATABASE IF EXISTS database_123456789abcdef"
```
Comment (antonio2368) [inline tests/queries/0_stateless/02775_show_columns_mysql_compatibility.sh:18]: `no-parallel` can be removed and `${CLICKHOUSE_DATABASE}` can be used to get the unique generated database for this test.
